django/db/models/expressions.py

```diff
@@ -600,7 +601,10 @@ def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize
         c = self.copy()
         c.is_summary = summarize
         for pos, arg in enumerate(c.source_expressions):
-            c.source_expressions[pos] = arg.resolve_expression(query, allow_joins, reuse, summarize, for_save)
+            resolved = arg.resolve_expression(query, allow_joins, reuse, summarize, for_save)
+            c.source_expressions[pos] = resolved
+            c.filterable = c.filterable and resolved.filterable
+
```

Comment:
I don't think that we should change anything in the `resolve_expression()`, IMO the easiest way to implement this is to override `filterable` property for `CombinedExpression` and `Func`, e.g.
```diff
--- a/django/db/models/expressions.py
+++ b/django/db/models/expressions.py
@@ -454,6 +454,10 @@ class CombinedExpression(SQLiteNumericMixin, Expression):
         c.rhs = c.rhs.resolve_expression(query, allow_joins, reuse, summarize, for_save)
         return c
 
+    @property
+    def filterable(self):
+        return self.lhs.filterable and self.rhs.filterable
+
 
 class DurationExpression(CombinedExpression):
     def compile(self, side, compiler, connection):
@@ -630,6 +634,10 @@ class Func(SQLiteNumericMixin, Expression):
         copy.extra = self.extra.copy()
         return copy
 
+    @property
+    def filterable(self):
+        return all(expr.filterable for expr in self.get_source_expressions())
+
```

Reply:
Ok, that's funny:
`return all(expr.filterable for expr in self.get_source_expressions())`
kinda works, but 
`return all(list(expr.filterable for expr in self.get_source_expressions()))`
doesn't. 

Consider the following example
```
Product.objects.annotate(
    row=Window(expression=RowNumber()),
    total=Sum('row', filter=Q(id=1))
).filter(
    total=1
)
```

The problem here is that expr can be `WhereNode` that doesn't have `filterable`. When we refer it, AttributeError is raised and then swallowed by `getattr` in `build_filter`, thus the expression is considered filterable. 

But because `Aggregate.get_source_expressions` returns `WhereNode` in the last position and `all` is lazy, etherything is fine - either the expression is determined to be non filterable before reaching WhereNode, or it is indeed filterable.